docs/design-documents/platform/clocks/clocks.md

```diff
@@ -0,0 +1,418 @@
+
+```C++
+#include <chrono> 
+using namespace std::chrono_literals;
+auto delay = 5ms;
+```
+
+All Chrono APIs are in the namespace `std::chrono`, so it can be convenient to have a directive to import this complete namespace to reduce typing:
+
+```C++
+#include <chrono>
+using namespace std::chrono; // provides both types like std::chrono::microseconds and the chrono_literals
+
+microseconds t = 1h;
+printf("1 hour is %d us\n", int(t.count()));
+```
+
+Namespace directives should only be used in source files to avoid unexpected name collision for users. Header files intended for general use will need to fully qualify their use of Chrono with `std::chrono` prefixes. This also precludes use of literals:
+
+```C++
+// In a header file
+#include <chrono>
+void my_function(std::chrono::microseconds timeout = std::chrono::seconds{1});
+```
+
+### Using the RTOS clock
+
+**Basic delays**
+
+```C++
+for (;;) {
+    LED1 = !LED1;
+    ThisThread::sleep_for(500ms);
+}
+```
+
+**Prefer absolute time**
+
+Using absolute time aids precision, guaranteeing that there is no drift due to code execution time. This applies to all clocks.
+
+```C++
+auto next_wake = Kernel::Clock::now();
+for (;;) {
+    LED1 = !LED1;
+    next_wake += 500ms;
+    ThisThread::sleep_until(next_wake);
+}
+```
+
+Periodic callback APIs such as `Ticker` or `EventQueue::call_every` use this mechanism. If manually rescheduling events yourself, you should use the same technique, rather than using `call_in` within a callback.
+
+**Measuring elapsed time via the RTOS**
+
+```C++
+auto start = Kernel::Clock::now(); // type will be Kernel::Clock::time_point
+do_operation();
+milliseconds elapsed_time = Kernel::Clock::now() - start;
+printf("elapsed time = %d ms\n", int(elapsed_time.count()));
+```
+
+Explicitly stating that `elapsed_time` is `milliseconds` provides a cross-check that the `count()` for the print really is milliseconds. If the clock had lower resolution, it would be implicitly converted to milliseconds. If the clock had higher resolution, it would be a compilation error - see "being more generic" below to handle this.
+
+### Using the high-resolution clock
+
+**Measuring elapsed time using Timer**
+```C++
+template<class F>
+microseconds time_operation(const F &operation)
+{
+    Timer timer;
+    timer.start(); // deep sleep is automatically locked while timer is running
+    operation();
+    return timer.elapsed_time();
+} // Deep sleep automatically unlocked when timer is destroyed.
+
+microseconds print_time = time_operation([] { printf("Hello!\n"); });
+printf("Printing that took %d us\n", int(print_time.count()));
+```
+
+This example relies on the destruction of the timer to unlock the clock. If using a timer that doesn't get destroyed, you must remember to stop it manually to allow deep sleep to be entered again.
```

Comment:
Yes and no. The clocks have a locking abstraction, as is shown in the next example. If you're using the clock, then you can lock it to make sure it keeps running. For the `HighResClock`, that then locks deep sleep. So it's a 1 step indirect abstraction, and that's what I was thinking of.

However, `Timer` isn't actually implemented on top of a `Clock` (although it could be) - it is actually locking deep sleep directly. It's got a base doing run-time polymorphism as `TimerBase(ticker_data_t)` - it could have been templates like `TimerBase<Clock>`. The first form was minimal structural change, and would lead to smaller code size if using both `Timer` and `LowPowerTimer. The second would be smaller if using only one.

But it's doing the deep sleep lock for the same logical reason to "lock" the clock represented by the `ticker_data_t` to keep running. And it's a contrast to the manual clock locks in the next example.

Reply:
I know I've already asked a lot of questions about that and I'm sorry if I'm a bit slow to get it. Coming back to the documentation after a few weeks, I'm still a little confused.

Here you say(emphasis mine)

>  If you're using the clock, then **you can lock it to make sure it keeps running**.

Which I think I understand correctly.

And [here](https://github.com/ARMmbed/mbed-os/pull/12425/files#diff-64f10ab2be6a2b6f14ac228129e91028R47) you say:

> [high-resolution clock] It does not run in deep sleep, so its users will usually **lock deep sleep**.

In this case, do you **lock deep sleep**:
- to use it (i.e. go into deep sleep) and *make sure it keeps sleeping* even with the high-res clock (that will stop running)
- or actually to **not go** into deep sleep so that your high-res clock can still run?

My guess is the second, but I'd say something like

> you can lock the high resolution clock to disable deep sleep and make sure the ticker keeps running.

Sorry again for getting back at it again :)

Reply:
> not go into deep sleep so that your high-res clock can still run?

This.

Our terminology around deep sleep locks is backwards. The "lock deep sleep" calls lock out deep sleep from being entered. The high-res clock will stop if deep sleep is entered.

(Android got this right by calling the lock that stops suspend from being entered a "wake lock", not a "suspend lock".)

In computer science lock something while using it, and unlock it when finished. (I guess the metaphor is a toilet, not a car). You then have to figure out what to call that lock. If we'd called it a "shallow sleep lock", things may have been better.

Reply:
Thanks @kjbracey-arm it makes sense, again ;)

Wouldn't mbed os 6.0 be a good time to also change the name of the deep sleep lock to shallow sleep lock? To make it clearer for new comers? Or is this intended for expert users only?